typings/index.d.ts

```diff
@@ -0,0 +1,737 @@
+export as namespace dyo
+
+declare namespace dyo {
+	type Key = string | number | any
+	type Type = Component | PromiseLike<Node> | string | null
+	type Target = Element | Document | ShadowRoot | DocumentFragment
+
+	type Node<Props = any> = Interface<Props> | object | string | number | boolean | null | undefined | void
+	type Collection = Node[] | Node
+	type Properties<Props, RefType = any> = Readonly<Props & Attributes & {children?: Collection; ref?: Ref<RefType>, [props: string]: any}>
+
+	type Ref<Value> = RefObject<Value> | RefCallback<Value>
+	type RefObject<Value> = {current: Value | null }
+	type RefCallback<Value> = (value: Value | null) => void
+
+	type SetStateAction<State> = State | ((state: State) => State)
+	type Dispatch<Action> = (value: Action) => void
+
+	type Reducer<State, Action> = (state: State, action: Action) => State
+	type ReducerState<Reduce extends Reducer<any, any>> = Reduce extends Reducer<infer State, any> ? State : never
+	type ReducerAction<Reduce extends Reducer<any, any>> = Reduce extends Reducer<any, infer Action> ? Action : never
+
+	type EffectCallback<Dependency, Props> = (deps?: Dependency, props?: Props) => (void | (() => void | undefined))
+	type MemoCallback<Dependency, Type> = (deps?: Dependency) => Type
+
+	type DependencyList = ReadonlyArray<any>
+
+	interface Interface<Props = {}> {type: Type, props: Props & {children: Collection}, ref: Ref<any> | null, key: Key}
+	interface Attributes {key?: Key}
+	interface Handler<Event, Props = {}> {(event: Event, props?: Props): any}
+	interface Exception<Value> {name: string, type: string, stack: string, message: Value}
+	interface Component<Props = {}> { (props: Properties<Props>): Node }
+	interface Context<Value> extends Component<{value: Value, children: Collection}> {}
+
+	function useRef<Ref, Props = {}> (value: Ref | ((props: Props) => Ref)): RefObject<Ref>;
+	function useState<State, Props = {}> (state: State | ((props: Props) => State)): [State, Dispatch<SetStateAction<State>>]
+	function useReducer<Reduce extends Reducer<State, any>, State = any> (reducer: Reduce, state: State & ReducerState<Reduce>): [ReducerState<Reduce>, Dispatch<ReducerAction<Reduce>>]
+
+	function useCallback<Value extends (...args: any[]) => any> (callback: Value): Value
+	function useMemo<Dependency extends DependencyList, Value = any> (factory: MemoCallback<Dependency, Value>, deps?: Dependency): Value
+
+	function useLayout<Dependency extends DependencyList, Props = {}> (callback: EffectCallback<Dependency, Props>, deps?: Dependency): void
+	function useEffect<Dependency extends DependencyList, Props = {}> (callback: EffectCallback<Dependency, Props>, deps?: Dependency): void
+
+	function useResource<Value, Props = {}> (callback: (props: Props) => PromiseLike<Value> | PromiseLike<Response>): Value
+	function useContext<Value>(context: Context<Value>): Value
+	// function useContext<Value extends Component>(context: Value): Pick<Pick<ReturnType<Value>, 'props'>, 'value'>
```

Comment:
Trying to type this so that "string" in the following picks it from the props value of the Context returned from `Provider`.

```js
function Provider({value, children}: {value: string, children: any}) {
  return h(Context, {value: 'string'}, children)
}

const string = useContext(Provider)
```

Reply:
The former mealy takes it from the Providers props `{value <- this, children}`.

Reply:
I think i've got this solved. Turns out `Pick` doesn't do what i thought it did and accessing type properties in TypeScript is done with `type['key']` instead of `type.key`.